provider/apply.go

```diff
@@ -270,6 +273,46 @@ func (s *RawProviderServer) ApplyResourceChange(ctx context.Context, req *tfprot
 			return resp, nil
 		}
 
+		// wait for the resource to be deleted
+		s.logger.Trace("[ApplyResourceChange][Apply]", "Waiting for resource to be deleted...")
+		timeout := defaultDeleteTimeout
+		if v, ok := priorStateVal["timeouts"]; ok && !v.IsNull() {
+			timeouts := map[string]tftypes.Value{}
+			v.As(&timeouts)
+			if v, ok := timeouts["delete"]; ok && !v.IsNull() {
+				v.As(&timeout)
+			}
+		}
+		s.logger.Trace("[ApplyResourceChange][Apply]", "Using timeout %q for delete", timeout)
+		duration, err := time.ParseDuration(timeout)
+		if err != nil {
+			resp.Diagnostics = append(resp.Diagnostics,
+				&tfprotov5.Diagnostic{
+					Severity: tfprotov5.DiagnosticSeverityError,
+					Detail:   err.Error(),
+					Summary:  fmt.Sprintf("error parsing timeout %q: %v", timeout, err),
+				})
+			return resp, nil
+		}
+		end := time.Now().Add(duration)
+		for {
+			_, err := rs.Get(ctx, rname, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				s.logger.Trace("[ApplyResourceChange][Apply]", "Resource is deleted")
+				break
+			}
+			time.Sleep(1 * time.Second)
+			if time.Now().After(end) {
+				resp.Diagnostics = append(resp.Diagnostics,
+					&tfprotov5.Diagnostic{
```

Comment:
I thought it made sense to make this a warning rather than an error that terminates the apply. We just want to action the user to go and investigate manually.

Reply:
I think the delete needs to fail here, otherwise the resources is removed from state and orphaned by Terraform.

Here's a test run using the default timeout of 1m (no timeout specified) on a resource with a finalizer (so it will never delete):

```
$ terraform destroy --auto-approve

Plan: 0 to add, 0 to change, 1 to destroy.
kubernetes_manifest.consul: Destroying...
kubernetes_manifest.consul: Still destroying... [10s elapsed]
kubernetes_manifest.consul: Still destroying... [20s elapsed]
kubernetes_manifest.consul: Still destroying... [30s elapsed]
kubernetes_manifest.consul: Still destroying... [40s elapsed]
kubernetes_manifest.consul: Still destroying... [50s elapsed]
kubernetes_manifest.consul: Still destroying... [1m0s elapsed]
kubernetes_manifest.consul: Destruction complete after 1m1s

Destroy complete! Resources: 1 destroyed.

$ terraform destroy --auto-approve
No changes. No objects need to be destroyed.

Either you have not created any objects yet or the existing objects were already deleted outside of Terraform.

Destroy complete! Resources: 0 destroyed.

$ kubectl get service tf-k8s-test-consul-server
NAME                        TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)             AGE
tf-k8s-test-consul-server   ClusterIP   None         <none>        8500/TCP,8600/TCP   5m26s

$ terraform show
$
```

Any resources which would normally take more than a minute to destroy are now unmanaged by Terraform, and could still exist on the cluster.